apps/public/public.go

```diff
@@ -0,0 +1,250 @@
+func WithCache(accessor cache.ExportReplace) Option {
+	return func(o *Options) {
+		o.Accessor = accessor
+	}
+}
+
+// Client is a representation of authentication client for public applications as defined in the
+// package doc. For more information, visit https://docs.microsoft.com/azure/active-directory/develop/msal-client-applications.
+type Client struct {
+	base.Client
+}
+
+// New is the constructor for Client.
+func New(clientID string, options ...Option) (Client, error) {
+	opts := Options{Authority: base.AuthorityPublicCloud}
+
+	for _, o := range options {
+		o(&opts)
+	}
+	if err := opts.validate(); err != nil {
+		return Client{}, err
+	}
+
+	base, err := base.New(clientID, opts.Authority, opts.Accessor, requests.NewToken())
+	if err != nil {
+		return Client{}, err
+	}
+	return Client{base}, nil
+}
+
+// CreateAuthCodeURL creates a URL used to acquire an authorization code.
+func (pca Client) CreateAuthCodeURL(ctx context.Context, clientID, redirectURI string, scopes []string) (string, error) {
+	return pca.Client.AuthCodeURL(ctx, clientID, redirectURI, scopes, pca.AuthParams)
+}
+
+// AcquireTokenSilentOptions are all the optional settings to an AcquireTokenSilent() call.
+// These are set by using various AcquireTokenSilentOption functions.
+type AcquireTokenSilentOptions struct {
+	// Account represents the account to use. To set, use the WithSilentAccount() option.
+	Account Account
+}
+
+// AcquireTokenSilentOption changes options inside AcquireTokenSilentOptions used in .AcquireTokenSilent().
+type AcquireTokenSilentOption func(a *AcquireTokenSilentOptions)
+
+// WithSilentAccount uses the passed account during an AcquireTokenSilent() call.
+func WithSilentAccount(account Account) AcquireTokenSilentOption {
+	return func(a *AcquireTokenSilentOptions) {
+		a.Account = account
+	}
+}
+
+// AcquireTokenSilent acquires a token from either the cache or using a refresh token.
+func (pca Client) AcquireTokenSilent(ctx context.Context, scopes []string, options ...AcquireTokenSilentOption) (AuthenticationResult, error) {
+	opts := AcquireTokenSilentOptions{}
+	for _, o := range options {
+		o(&opts)
+	}
+
+	silentParameters := base.AcquireTokenSilentParameters{
+		Scopes:      scopes,
+		Account:     opts.Account,
+		RequestType: accesstokens.RefreshTokenPublic,
+	}
+
+	return pca.Client.AcquireTokenSilent(ctx, silentParameters)
+}
+
+// AcquireTokenByUsernamePassword acquires a security token from the authority, via Username/Password Authentication.
+// NOTE: this flow is NOT recommended.
+func (pca Client) AcquireTokenByUsernamePassword(ctx context.Context, scopes []string, username string, password string) (AuthenticationResult, error) {
+	authParams := pca.AuthParams
+	authParams.Scopes = scopes
+	authParams.AuthorizationType = authority.AuthorizationTypeUsernamePassword
+	authParams.Username = username
+	authParams.Password = password
+
+	token, err := pca.Client.Token.UsernamePassword(ctx, authParams)
```

Comment:
>Token [](start = 26, length = 5)

that the requests/flows internally are "hanging" on Token (applies everywhere), seems odd to me.

Reply:
Not sure I understand (unless it is the name?)  keegan and I have been talking about reversing the name (this is an old package name) from requests.Token to token.Requestor.  

This package is the upper layer logic that may make 1 or more REST calls in order to get a token from various sources.

Does that help, am I misunderstanding?

Reply:
I think it is the name, this isn't obvious from the name that "Token" is a token sub-client/token requestor.

Reply:
it's the name ... token is not the thing that does the flow, it's rather an artifact. this logic seems inversed to me. the request object is used to get tokens back in the response (among other things)